flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java

```diff
@@ -72,354 +62,237 @@
-
-		log.info("Recovered {} containers from previous attempts ({}).", containersFromPreviousAttempts.size(), containersFromPreviousAttempts);
-
-		for (final Container container : containersFromPreviousAttempts) {
-			final ResourceID resourceID = getContainerResourceId(container);
-			workerNodeMap.put(resourceID, new YarnWorkerNode(container, resourceID));
-		}
-	}
-
-	private void updateMatchingStrategy(final RegisterApplicationMasterResponse registerApplicationMasterResponse) {
-		final Optional<Set<String>> schedulerResourceTypesOptional =
-			registerApplicationMasterResponseReflector.getSchedulerResourceTypeNames(registerApplicationMasterResponse);
-
-		final WorkerSpecContainerResourceAdapter.MatchingStrategy strategy;
-		if (schedulerResourceTypesOptional.isPresent()) {
-			Set<String> types = schedulerResourceTypesOptional.get();
-			log.info("Register application master response contains scheduler resource types: {}.", types);
-			matchingStrategy = types.contains("CPU") ?
-				WorkerSpecContainerResourceAdapter.MatchingStrategy.MATCH_VCORE :
-				WorkerSpecContainerResourceAdapter.MatchingStrategy.IGNORE_VCORE;
-		} else {
-			log.info("Register application master response does not contain scheduler resource types, use '{}'.",
-				YarnConfigOptionsInternal.MATCH_CONTAINER_VCORES.key());
-		}
-		log.info("Container matching strategy: {}.", matchingStrategy);
-	}
-
-	protected NMClientAsync createAndStartNodeManagerClient(YarnConfiguration yarnConfiguration) {
-		// create the client to communicate with the node managers
-		NMClientAsync nodeManagerClient = NMClientAsync.createNMClientAsync(this);
-		nodeManagerClient.init(yarnConfiguration);
-		nodeManagerClient.start();
-		return nodeManagerClient;
-	}
-
-	@Override
-	protected Configuration loadClientConfiguration() {
-		return GlobalConfiguration.loadConfiguration(env.get(ApplicationConstants.Environment.PWD.key()));
-	}
+	// ------------------------------------------------------------------------
+	//  ResourceManagerDriver
+	// ------------------------------------------------------------------------
 
 	@Override
-	protected void initialize() throws ResourceManagerException {
+	protected void initializeInternal() throws Exception {
+		final YarnContainerEventHandler yarnContainerEventHandler = new YarnContainerEventHandler();
 		try {
-			resourceManagerClient = createAndStartResourceManagerClient(
-				yarnConfig,
+			resourceManagerClient = yarnResourceManagerClientFactory.createResourceManagerClient(
 				yarnHeartbeatIntervalMillis,
-				webInterfaceUrl);
+				yarnContainerEventHandler);
+			resourceManagerClient.init(yarnConfig);
+			resourceManagerClient.start();
+
+			final RegisterApplicationMasterResponse registerApplicationMasterResponse = registerApplicationMaster();
+			getContainersFromPreviousAttempts(registerApplicationMasterResponse);
+			updateMatchingStrategy(registerApplicationMasterResponse);
 		} catch (Exception e) {
```

Comment:
From my understanding, create and start the client would not throw `Exceptions`. We actually catch the exception thrown by `registerApplicationMaster`.

Reply:
I see. I just realized that the `ResourceManagerException` might have been misused here since we're catching exceptions related to YARN's resource manager. `ResourceManagerException` relates to Flink's resource manager though. Am I getting that right? It's unfortunate that there is this name clash.

Reply:
> I see. I just realized that the `ResourceManagerException` might have been misused here since we're catching exceptions related to YARN's resource manager. `ResourceManagerException` relates to Flink's resource manager though. Am I getting that right? It's unfortunate that there is this name clash.

Or thinking about it again. Probably, it's alright to use `ResourceManagerException` since it's still part of the `ResourceManager`'s context. 🤔 Never mind...